### ISSUE TITLE
PoseCov3D (C3D backbone)  reproducing results

### ISSUE THREAD

**Issue by ahmed-nady** [kennymckormick/pyskl#95]
I have a question regarding the C3D architecture, i noticed there is a difference between the architecture that you reported in your paper and the one exists here in your repository (the difference is using average pooling instead of carpooling and adding pool layer at stage 4). So, the accuracy that you reported based on architecture in paper or in this repository?    

Also, i tried to reproduce your results (C3D backbone) based on your code here(c3d.py) in this repository and the results :
top1_acc:	0.9119
top5_acc:	0.9944
mean_acc:0.9118
and the top1_acc that reported in your paper is 92.9 %?

If it is possible, could you which architecture you used to report your results (paper or c3d.py) and how to reach your reported results?
Thanks in advance.

**Comment by kennymckormick** [comment]
Are you using this config to reproduce the C3D results: https://github.com/kennymckormick/pyskl/blob/main/configs/posec3d/c3d_light_ntu60_xsub/joint.py? 
If so, I will retrain this config for a double check. 
Besides, I think that's a typo in our appendix. Actually, we are using average pooling in our PoseConv3D (C3D backbone). I will fix it later. There should not be a pooling layer before stage 4, but since I only released the C3D-light backbone, I just didn't modify that part.

**Comment by ahmed-nady** [comment]
Thanks @kennymckormick for your swift response. Because i used the mmaction2 repository before,  I also used it to train the PoseCon3D model using c3d.py (that file from your repository "pyskl")  and the config that i used is the same with  the one you pointed to [https://github.com/kennymckormick/pyskl/blob/main/configs/posec3d/c3d_light_ntu60_xsub/joint.py](url)
Also, i attached it for your reference.

`model = dict(
    type='Recognizer3D',
    backbone=dict(
        type='C3D',
        in_channels=17,
        base_channels=32,
        num_stages=3,
        temporal_downsample=False),
    cls_head=dict(
        type='I3DHead',
        in_channels=256,
        num_classes=60,
        spatial_type='avg',
        dropout_ratio=0.5),
    test_cfg=dict(average_clips='prob'))
gpu_ids =(0,1)
dataset_type = 'PoseDataset'
ann_file_train = '/mmaction2/Action Recognition/ntu60_xsub_train.pkl'  
ann_file_val = '/mmaction2/Action Recognition/ntu60_xsub_val.pkl'


left_kp = [1, 3, 5, 7, 9, 11, 13, 15]
right_kp = [2, 4, 6, 8, 10, 12, 14, 16]
train_pipeline = [
    dict(type='UniformSampleFrames', clip_len=48),
    dict(type='PoseDecode'),
    dict(type='PoseCompact', hw_ratio=1., allow_imgpad=True),
    dict(type='Resize', scale=(-1, 64)),
    dict(type='RandomResizedCrop', area_range=(0.56, 1.0)),
    dict(type='Resize', scale=(56, 56), keep_ratio=False),
    dict(type='Flip', flip_ratio=0.5, left_kp=left_kp, right_kp=right_kp),
    dict(
        type='GeneratePoseTarget',
        sigma=0.6,
        use_score=True,
        with_kp=True,
        with_limb=False),
    dict(type='FormatShape', input_format='NCTHW'),
    dict(type='Collect', keys=['imgs', 'label'], meta_keys=[]),
    dict(type='ToTensor', keys=['imgs', 'label'])
]
val_pipeline = [
    dict(type='UniformSampleFrames', clip_len=48, num_clips=1, test_mode=True),
    dict(type='PoseDecode'),
    dict(type='PoseCompact', hw_ratio=1., allow_imgpad=True),
    dict(type='Resize', scale=(-1, 64)),
    dict(type='CenterCrop', crop_size=64),
    dict(
        type='GeneratePoseTarget',
        sigma=0.6,
        use_score=True,
        with_kp=True,
        with_limb=False),
    dict(type='FormatShape', input_format='NCTHW'),
    dict(type='Collect', keys=['imgs', 'label'], meta_keys=[]),
    dict(type='ToTensor', keys=['imgs'])
]
test_pipeline = [
    dict(
        type='UniformSampleFrames', clip_len=48, num_clips=1, test_mode=True),
    dict(type='PoseDecode'),
    dict(type='PoseCompact', hw_ratio=1., allow_imgpad=True),
    dict(type='Resize', scale=(-1, 64)),
    dict(type='CenterCrop', crop_size=64),
    dict(
        type='GeneratePoseTarget',
        sigma=0.6,
        use_score=True,
        with_kp=True,
        with_limb=False,
        double=True,
        left_kp=left_kp,
        right_kp=right_kp),
    dict(type='FormatShape', input_format='NCTHW'),
    dict(type='Collect', keys=['imgs', 'label'], meta_keys=[]),
    dict(type='ToTensor', keys=['imgs'])
]
data = dict(
    videos_per_gpu=16,
    workers_per_gpu=4,
    test_dataloader=dict(videos_per_gpu=1),
    train=dict(
        type=dataset_type,
        ann_file=ann_file_train,
        data_prefix='',
        class_prob={i: 1 for i in range(22)},
        pipeline=train_pipeline),
    val=dict(
        type=dataset_type,
        ann_file=ann_file_val,
        data_prefix='',
        pipeline=val_pipeline),
    test=dict(
        type=dataset_type,
        ann_file=ann_file_val,
        data_prefix='',
        pipeline=test_pipeline))

# optimizer
optimizer = dict(
    type='SGD', lr=0.001, momentum=0.9,
    weight_decay=0.0003)  # this lr is used for 8 gpus
optimizer_config = dict(grad_clip=dict(max_norm=40, norm_type=2))
# learning policy
lr_config = dict(policy='CosineAnnealing', by_epoch=False, min_lr=0)
total_epochs = 240
checkpoint_config = dict(interval=5)
workflow = [('train', 5)]
evaluation = dict(
    interval=5,
    metrics=['top_k_accuracy', 'mean_class_accuracy'],
    topk=(1, 5))
log_config = dict(
    interval=20, hooks=[
        dict(type='TextLoggerHook'),
        dict(type='TensorboardLoggerHook'),
    ])
dist_params = dict(backend='nccl')

log_level = 'INFO'
work_dir = './work_dirs/posec3d/c3d_light_ntu60_xsub/'
load_from = None
resume_from = None
find_unused_parameters = False`

**Comment by ahmed-nady** [comment]
@kennymckormick  
Also, i set the learning rate to 0.001 because i used two GPUs. Is the learning rate should be 0.005 for 2 gpus and 16 videos/gpu? and what you think about the effect of this setting for LR on the accuracy?

**Comment by kennymckormick** [comment]
I'm afraid the learning rate you are using might be too small. 
In our reference config: https://github.com/kennymckormick/pyskl/blob/main/configs/posec3d/c3d_light_ntu60_xsub/joint.py, we set the initial learning rate to 0.4 when the total batch size is 256. Thus I think you should set the lr to 0.05, when you are using batchsize 32.

**Comment by ahmed-nady** [comment]
@kennymckormick  Thanks for providing me with a tip for setting LR. The accuracy is improved but still has not reached the reported accuracy. So, i am asking about:
1-the setting for weight_decay, is it changed according to batch size or it depends on the number of epochs (changes happen at beginning of each epoch, not the iteration as you know)?
2-The top 1 accuracy is calucated based on 1-video testing or 10-video testing?
3- For training the network with both joints and limbs, do you train the network with joints as input and train another network with limbs and then fuse the score by average?

**Comment by kennymckormick** [comment]
1. The weight decay does not change according to the batch size, it is always fixed. 
2. The top-1 acc is based-on 10-clip testing. 
3. Yes.
